### PR TITLE
Fix GRPCListener initialization race condition

### PR DESCRIPTION
**Pull request by ShahabT**
## What changed?

Fix race condition in `RPCFactory.GetGRPCListener` which can lead to nil return value.

## Why?

Creates flakiness in tests. For example it caused "nil pointer dereference error" in [here](https://github.com/temporalio/temporal/actions/runs/9506317543/job/26203365793?pr=6094).

## How did you test it?

None.

## Potential risks

None.

## Documentation

None.

## Is hotfix candidate?

No.

### CODE REVIEW

**Review by dnr — APPROVED** [review]
OnceValue is definitely nicer but I don't see how the previous code can fail, where's the race?

File: common/rpc/rpc.go [hunk 1]
```diff
@@ -162,19 +162,20 @@ func (d *RPCFactory) GetInternodeClientTlsConfig() (*tls.Config, error) {
...
+	var err error
+	grpcListener, err := net.Listen("tcp", hostAddress)
 
-		d.logger.Info("Created gRPC listener", tag.Service(d.serviceName), tag.Address(hostAddress))
-	})
+	if err != nil {
+		d.logger.Fatal("Failed to start gRPC listener", tag.Error(err), tag.Service(d.serviceName), tag.Address(hostAddress))
+	}
 
-	return d.grpcListener
```
Comment (ShahabT) [inline common/rpc/rpc.go:177]: @dnr I was thinking `d.grpcListener` may not synchronize between the goroutine that does the assignment in line 168 and another one which just waits for `Do` without running the func. In that case, the second goroutine may return nil. Is that right?

Comment (dnr) [reply common/rpc/rpc.go:177]: I thought that was maybe possible, but if you look at the implementation of Once, there should always be synchronization there, either Load observes the Store, or the mutex Lock observes the Unlock.

Actually even if you don't look at implementation, anything that Once could do to implement its guarantee "no call to Do returns until the one call to f returns" must synchronize somehow.

And if you look at OnceValue, it's just wrapping this pattern.

Comment (ShahabT) [reply common/rpc/rpc.go:177]: Makes sense.

Not sure how to explain the error though unless net.Listen returned `nil, nil`, or returned a listener with nil `Addr()`. I'm going to add those checks.

**Review by dnr — APPROVED** [review]
I can't explain how it ended up nil either, but we should merge this change anyway

**Pull request merged**